core/templates/dev/head/pages/topic_viewer/StoriesListDirective.js

```diff
@@ -0,0 +1,32 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for the stories.
+ */
+oppia.directive('storiesList', [
+  'UrlInterpolationService', function(UrlInterpolationService) {
+    return {
+      restrict: 'E',
+      scope: {
+        getTopicId: '&topicId',
+        canonicalStories: '=canonicalStoryList',
```

Comment:
Any specific reason for using '='? 
I think you can do `getCanonicalStories: '&canonicalStoriesList'` here as well.

Reply:
I am passing a list that's why I've used ```=```. Using ```&``` doesn't serve the purpose. I tried.

Reply:
The data type shouldn't matter. If you want one-way binding (and aren't modifying that list from within this directive) then `&` should work. Maybe you can explain a bit more what you tried and what went wrong?

If on the other hand you are modifying the passed-in stuff from within this directive, then '=' makes sense.

Reply:
Hi Sean, I did start with using ```&``` but only three (if total number is greater than three) tiles showed, also upon resizing I found that the tiles didn't render for mobile view. Though there was just one backend API call done. One possible explanation I found was [this](https://stackoverflow.com/questions/14695792/binding-array-to-directive-variable-in-angularjs).

Reply:
That's odd, I wouldn't expect that. Can you introspect canonicalStories by putting `<[canonicalStories]>` in the directive html and then seeing if that variable changes as you'd expect?

Also, if you're not able to figure out what's going on, can you give full repro instructions? Maybe Akshay or I can take a look.

Thanks!

Reply:
Hi! I tried printing the ```canonicalStories``` but the variable didn't change. I'm not sure what's causing this issue. PTAL!

Reply:
You have confirmed that, by following the same instructions, different behaviour occurs with `=` and `getCanonicalStories: '&canonicalStoriesList'`?

Can you give a full list of steps, starting from your current branch, for how to reproduce the issue with '&'? Include every step, starting from "change `=` to `getCanonicalStories: '&canonicalStoriesList'`, and any code changes needed (you can talk in terms of find/replace), and then what manual testing steps you did after that.

I'm going to bed now, but @aks681 might be able to take a look. If not, I can try to help debug tomorrow, too.

Reply:
Yes! I confirm that there are different behaviors for ```=``` and ```&```. 

Steps to repro:
- Change ```=canonicalStoryList``` to ```&canonicalStoryList``` on line 24 in StoryListDirective.js.
- Add ```()``` to canonicalStories on line 1, 19 and 40 in story_list_directive.html.
- Refresh the topic viewer page.

You won't be able to find scroll right button on the carousel.
EDIT: The story list is visible in the mobile view. It missed it adding ```()``` on line 40.

Reply:
When adding () to canonicalStories, did you change it to getCanonicalStories?

Also, it seems like (from your edit) at least one of the bugs is fixed. Could you confirm which bugs remain and explain how to reproduce them?

Reply:
Yes, I did.

Now the problem that occurs is only 3 tiles are showing and the next button gets hidden.

Reply:
I dug into this a bit.

- I'm a bit confused about initCarousels(). Why does it fetch data from the topic viewer backend API service? Shouldn't it be using the data in the canonical stories list that's passed in? Otherwise you have two sources of truth here.
- Leave your HTML file as-is, but after the directive has been initialized, write something like `$scope.canonicalStories = $scope.getCanonicalStories();`. That will keep the flow of information from the parent one-way, and you can still bind to a canonicalStories variable that's correct and that won't change.

Does that help?

Reply:
Hi @seanlip, Sorry for the late reply, I missed your comment. 
- I fetched data from the API service because the passed data was displaying ```undefined``` but was reusable in the HTML. I am not sure why that was happening so I changed the service to return topicData if it was fetched earlier.
- I tried this but surprisingly none of the tiles is appearing now. 

I'm sure there's an explanation of this weird behaviour. I'm pushing the code with the changes you suggested and the recent minor mock changes.

Reply:
I'm mostly confused: the current code still looks like an "=" to me and there's nothing in the controller. Was that intentional?

Yes, you do have to think through the race conditions. Make sure to only read the data a single time, once things are fully-loaded. Don't make the data editable when it should not be -- that is not the right solution.